src/Http/MiddlewareQueue.php

```diff
@@ -24,62 +26,108 @@
 class MiddlewareQueue implements Countable
 {
     /**
-     * The queue of middleware callables.
+     * The queue of middlewares.
      *
      * @var array
      */
     protected $queue = [];
 
     /**
-     * Get the middleware object at the provided index.
+     * The queue of middleware callables.
+     *
+     * @var array
+     */
+    protected $callables = [];
+
+    /**
+     * Get the middleware at the provided index.
      *
      * @param int $index The index to fetch.
      * @return callable|null Either the callable middleware or null
      *   if the index is undefined.
      */
     public function get($index)
     {
-        if (isset($this->queue[$index])) {
-            return $this->queue[$index];
+        if (isset($this->callables[$index])) {
+            return $this->callables[$index];
         }
 
-        return null;
+        return $this->resolve($index);
+    }
+
+    /**
+     * Resolve middleware name to callable.
+     *
+     * @param int $index The index to fetch.
+     * @return callable|null Either the callable middleware or null
+     *   if the index is undefined.
+     */
+    protected function resolve($index)
+    {
+        if (!isset($this->queue[$index])) {
+            return null;
+        }
+
+        if (is_string($this->queue[$index])) {
+            $class = $this->queue[$index];
+            $className = App::className($class, 'Middleware', 'Middleware');
+            if (!$className || !class_exists($className)) {
+                throw new RuntimeException(sprintf(
+                    'Middleware "%s" was not found.',
+                    $class
+                ));
+            }
+
+            $callable = new $class;
+        } else {
+            $callable = $this->queue[$index];
+        }
+
+        return $this->callables[$index] = $callable;
     }
 
     /**
      * Append a middleware callable to the end of the queue.
      *
-     * @param callable $callable The middleware callable to append.
+     * @param callable|string|array $middleware The middleware(s) to append.
      * @return $this
      */
-    public function add(callable $callable)
+    public function add($middleware)
     {
-        $this->queue[] = $callable;
+        if (is_array($middleware)) {
+            return $this->queue = array_merge($this->queue, $middleware);
```

Comment:
This won't return `$this` which I think it should. No reason for this method to not be chainable.